### ISSUE TITLE
Avoid using `pb` internals for datastore on App Engine

### ISSUE THREAD

**Issue by dhermes** [googleapis/google-cloud-python#298]
See #282 for context.

From @pcostell:

> Just a random note about proto usage:
> If you plan to let users run gcloud-python in the App Engine python runtime, proto access is
> really slow (I believe because of the sandbox). I would limit proto access to serialization
> and deserialization only.

and

> Not 100% sure, but I think it uses a C proto library so it goes outside the process boundary.

**Comment by pcostell** [comment]
Awesome, good to know. Something that may be important: The reason GAE protos have slow access is that they use the C++ library, which requires a process boundary cross. This is used instead of a pure python library because serialization and deserialization is much faster in C++ than in python. In may be worthwhile to still minimize proto access in order to minimize SWIG-code entry points when users do have the C++ libraries (running on VMs or if GAE adds google.protobuf to the runtime).

**Comment by dhermes** [comment]
Luckily the `protobuf` [library][1] only uses C-code from the standard library. As @tseaver pointed out earlier today, `struct` is plenty fast.

I [analyzed][2] the imports, and the only imports from outside the library are:
```
StringIO.StringIO
cStringIO
cStringIO.StringIO
copy_reg
copyreg  # Py3, comment added by dhermes
io.BytesIO  # Py3, comment added by dhermes
operator
os
re
struct
sys
uuid
weakref
```

[1]: https://github.com/google/protobuf
[2]: https://gist.github.com/dhermes/7ffb1c3bc6ba554f86a0

**Comment by pcostell** [comment]
Currently, when using the externally available protobuf library on App Engine, it is using the python-only version of the protobuf library (it's impossible to upload a C library to the python runtime). However, once we are able to get the external C library into the App Engine runtime it will start using the C version instead of the python version. The C version has faster encode/decode times however you do pay a penalty when accessing protocol buffer fields because it has to cross a process boundary. We saw this in NDB's Key implementation which tried to store everything internally in protobuf rather than storing the fields as native python types.

I went through and it looks like you only access the protobuf on serialization and deserialization, which is perfect, so I'm leaving this as closed.
